# Portfolio Project: Online Retail Exploratory Data Analysis with Python

## Overview

In this project, you will step into the shoes of an entry-level data analyst at an online retail company, helping interpret real-world data to help make a key business decision.

## Case Study
In this project, you will be working with transactional data from an online retail store. The dataset contains information about customer purchases, including product details, quantities, prices, and timestamps. Your task is to explore and analyze this dataset to gain insights into the store's sales trends, customer behavior, and popular products. 

By conducting exploratory data analysis, you will identify patterns, outliers, and correlations in the data, allowing you to make data-driven decisions and recommendations to optimize the store's operations and improve customer satisfaction. Through visualizations and statistical analysis, you will uncover key trends, such as the busiest sales months, best-selling products, and the store's most valuable customers. Ultimately, this project aims to provide actionable insights that can drive strategic business decisions and enhance the store's overall performance in the competitive online retail market.

## Project Objectives
1. Describe data to answer key questions to uncover insights
2. Gain valuable insights that will help improve online retail performance
3. Provide analytic insights and data-driven recommendations

## Dataset

The dataset you will be working with is the "Online Retail" dataset. It contains transactional data of an online retail store from 2010 to 2011. The dataset is available as a .xlsx file named `Online Retail.xlsx`. This data file is already included in the Coursera Jupyter Notebook environment, however if you are working off-platform it can also be downloaded [here](https://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx).

The dataset contains the following columns:

- InvoiceNo: Invoice number of the transaction
- StockCode: Unique code of the product
- Description: Description of the product
- Quantity: Quantity of the product in the transaction
- InvoiceDate: Date and time of the transaction
- UnitPrice: Unit price of the product
- CustomerID: Unique identifier of the customer
- Country: Country where the transaction occurred

## Tasks

You may explore this dataset in any way you would like - however if you'd like some help getting started, here are a few ideas:

1. Load the dataset into a Pandas DataFrame and display the first few rows to get an overview of the data.
2. Perform data cleaning by handling missing values, if any, and removing any redundant or unnecessary columns.
3. Explore the basic statistics of the dataset, including measures of central tendency and dispersion.
4. Perform data visualization to gain insights into the dataset. Generate appropriate plots, such as histograms, scatter plots, or bar plots, to visualize different aspects of the data.
5. Analyze the sales trends over time. Identify the busiest months and days of the week in terms of sales.
6. Explore the top-selling products and countries based on the quantity sold.
7. Identify any outliers or anomalies in the dataset and discuss their potential impact on the analysis.
8. Draw conclusions and summarize your findings from the exploratory data analysis.

## Task 1: Load the Data

In [2]:
# Imports

import os
import sys
import pandas as pd

In [47]:
# Loading the Excel dataset into a Pandas DataFrame

retail_df = pd.read_excel('../Datasets/OnlineRetail.xlsx')
print(retail_df.head(5))

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  


In [4]:
# Exploring raw data before cleaning and validating

# 1. Top 10 Most Expensive Products Based on Unit Price
# Sort the dataframe by "UnitPrice" and display the desired products
most_expensive = retail_df.sort_values(by='UnitPrice', ascending=False).head(10)
print(most_expensive[['Description', 'UnitPrice']])

            Description  UnitPrice
222681           Manual   38970.00
524602       AMAZON FEE   17836.46
43702        AMAZON FEE   16888.02
43703        AMAZON FEE   16453.71
16356        AMAZON FEE   13541.33
15016        AMAZON FEE   13541.33
15017        AMAZON FEE   13541.33
16232        AMAZON FEE   13474.79
524601       AMAZON FEE   11586.50
299982  Adjust bad debt   11062.06


# Top 10 Most Expensive Products Based on Unit Price
After attempting to explore the raw data, there were many issues within the dataset that immediately arose. These issues being NaN values, product descriptions that were not actual products, and data values not having a consistent format.

As seen from the data above, the top 10 most expensive products based on their unit prices were not actual products. 
- One was 'Manual' which could be assumed to be just a manual input for the sale of a product. While it is unknown the product/reasoning as to why this was inputted, this would actually turn out to have the highest unit price out of all of the items. 
- The next is 'AMAZON FEE' which can be deduced to the company paying Amazon a certain percentage for having used Amazon's platform to sell one of their own products. This is further supported as the quantity is -1, which results in a negative value when multiplied by the unit price.
- And finally there is 'Adjust bad debt' which would be the company paying back a customer or client for a certain reason (whether it be a refund, reimbursement, courtesy payment, etc.)

## Methods Used
- .sort_values()
    - This sorts the values according to a certain parameter(s) set. In this case, it sorts the data by "UnitPrice" and sets ascending to false. Thus, resulting in the items being sorted from most expensive to least expensive.
- .head()
    - Stores only the first x amount of values. In this case, it is the first 10 values.

In [5]:
# 2. Top 4 Months With the Most Sales in 2011
# Begin by reformatting InvoiceDate to datetime, which allows for later usage of .dt functions
retail_df['InvoiceDate'] = pd.to_datetime(retail_df['InvoiceDate'])

# Filter for 2011 only, using .dt.year
retail_df_2011 = retail_df[retail_df['InvoiceDate'].dt.year == 2011]

# Filter 2011 only sales by each month and sum the quantities of the products sold
monthly_sales_2011 = (retail_df_2011
                      .groupby(retail_df_2011['InvoiceDate'].dt.month)['Quantity'] # Finding month and quantity of each product sold
                      .sum() # Adds up all the quantities of products in a month
                      .reset_index()) # Once grouping and summing are finished, this will use the index number as the column number to represent the month number being visualized

# Sort by quantity and print the desired amount of products
most_sales_2011 = monthly_sales_2011.sort_values(by='Quantity', ascending=False).head(4)
print(most_sales_2011)

    InvoiceDate  Quantity
10           11    740286
9            10    570532
8             9    549817
7             8    406199


# Top 4 Months With the Most Sales in 2011

In order to determine the top 4 months of 2011 with the most sales, it was important to consider the actual variable to scrutinize. Determining between quantity of product sold and the actual reported revenue. Considering that "sales" refers to action of selling a product and does not specify "top-grossing", then it would be accurate to consider the quantity instead of looking at the price. Not only this, but after seeing the top ten products sold were not even actual products, using the raw data for prices may not provide the most accurate results.

First, the dataset's "InvoiceDate" variable would need to be formatted differently so it could be read properly by the method for determining whether a sale was conducted in 2011 or not. Once filtered for sales conducted in 2011, the remaining data could now be sorted by the month they were conducted in. Once properly sorted, a new dataset would be made containing all of the months and their total quantities of products sold.

As seen from the output, the top 4 months with the most sales were August, September, October, and November in order from least to greatest.

## Methods Used
- .to_datetime()
    - Converts argument(s) to datetime format, which is essential when needing to parse strings (such as when triyng to find specific months of a specific year with the most sales).
- .dt.year()
    - Returns a Numpy array containing the year value from an object in the datetime format.
- .groupby()
    - Used to split the dataframe into groups based on a desired variable. In this case, splitting the dataframe by dates and extracting the quantities from each sale.
- .sum()
    - Adds up every quantity of product sold from the current month being grouped.
- .reset_index()
    - Since a new dataframe is being made, this essentially creates a new index for every month and its sale quantity being logged.

# Task 2: Perform Data Cleaning

In [48]:
# Finding missing values
print(retail_df.isnull().sum())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


In [49]:
# Identifying rows with missing values
print('Rows that have a missing description:\n', retail_df[retail_df['Description'].isnull()].head(3))

print('Rows that have a missing customer ID:\n', retail_df[retail_df['CustomerID'].isnull()].head(3))

Rows that have a missing description:
      InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
622     536414     22139         NaN        56 2010-12-01 11:52:00        0.0   
1970    536545     21134         NaN         1 2010-12-01 14:32:00        0.0   
1971    536546     22145         NaN         1 2010-12-01 14:33:00        0.0   

      CustomerID         Country  
622          NaN  United Kingdom  
1970         NaN  United Kingdom  
1971         NaN  United Kingdom  
Rows that have a missing customer ID:
      InvoiceNo StockCode                      Description  Quantity  \
622     536414     22139                              NaN        56   
1443    536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444    536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
622  2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443 2010-12-01 14:32:00       2.51     

In [53]:
# Finding rows with duplicate data
dupes = retail_df[retail_df.duplicated()]
print('There are', len(dupes), 'rows with duplicate data in this dataset:\n', dupes)

There are 5268 rows with duplicate data in this dataset:
        InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice  Cu

In [51]:
# Identifying rows that have a unit price <= 0
print(retail_df[retail_df['UnitPrice'] <= 0])

       InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                           NaN        56   
1970      536545     21134                           NaN         1   
1971      536546     22145                           NaN         1   
1972      536547     37509                           NaN         1   
1987      536549    85226A                           NaN         1   
...          ...       ...                           ...       ...   
536981    581234     72817                           NaN        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                           NaN        20   
538919    581422     23169                       smashed      -235   

               InvoiceDate  UnitPrice  CustomerID         Country  
622    2010-12-01 11:52:00        0.0         NaN  United Kingdom  
1970   2010-12-01 14:32

In [55]:
# Creating new DataFrame for cleaning raw data
# Removing rows with duplicate data
cleaned_retail_df = retail_df.drop_duplicates()

# Removing rows that are missing CustomerID
cleaned_retail_df.dropna(subset=['CustomerID'], inplace=True)

# Removing rows that have UnitPrice <= 0
cleaned_retail_df.drop(cleaned_retail_df[cleaned_retail_df['UnitPrice'] <= 0].index, inplace=True)

print(cleaned_retail_df.isnull().sum(), '\n')
print('Original Dataset:', len(retail_df), 'rows\nCleaned Dataset:', len(cleaned_retail_df), 'rows\n')
print((len(retail_df) - len(cleaned_retail_df)), 'rows have been deleted')

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64 

Original Dataset: 541909 rows
Cleaned Dataset: 401564 rows

140345 rows have been deleted


In order to conduct an accurate analysis on any given dataset, it is crucial to both clean and validate the raw data before any proper work can be done. There are multiple factors that must be addressed:

- Missing, incorrect, and duplicate data
- Fixing structural errors
- Uniform formatting across all data values
- Handling outliers

As a result of cleaning the dataset, rows with the following attributes have been deleted:

- All columns are exactly the same, indicating a duplicate row
- Unit prices less than or equal to zero
- Missing customer ID 

This has been done in order to remove duplicated data values that would cause more than 1 input of an event/sale that only occurred once. Rows with unit prices less than or equal to zero also serve no significance as they offer no descriptive information and only create more clutter to analyze. And since this is a customer-based analysis that is centralized on gaining insight on customer behavior, being unable to access which sale pertains to which customer is undesirable. Thus, rows with missing customer ID's must be deleted.

## Methods Used
- .isnull().sum()
    - Since .isnull() returns all rows that contain NaN values, this returns the total sum of rows containing NaN values grouped by each column
- .duplicated()
    - Provided insight into each duplicated row and the contents of them
- .drop_duplicates()
    - Drops all duplicate rows with exactly identical values and creates a new DataFrame with the resulting data
- .dropna(subset=['CustomerID'], inplace=True)
    - Drops all rows in the original DataFrame that have NaN values in "CustomerID"
- .drop(cleaned_retail_df[cleaned_retail_df['UnitPrice'] <= 0].index, inplace=True)
    - Finds the indices of rows with "UnitPrice" less than or equal to zero, drops them, and modifies the original DataFrame with this new data. 

As a result of cleaning, 140345 total rows were deleted.


# Task 3. Explore the Basic Statistics of the Dataset, Including Measures of Central Tendency and Dispersion.